x/auth/middleware/branch_store_test.go

```diff
@@ -0,0 +1,120 @@
+package middleware_test
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/cosmos/cosmos-sdk/simapp"
+	"github.com/cosmos/cosmos-sdk/testutil/testdata"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/cosmos/cosmos-sdk/x/auth/middleware"
+	minttypes "github.com/cosmos/cosmos-sdk/x/mint/types"
+)
+
+var blockMaxGas = uint64(simapp.DefaultConsensusParams.Block.MaxGas)
+
+func (s *MWTestSuite) TestBranchStore() {
+	testcases := []struct {
+		name         string
+		gasToConsume uint64 // gas to consume in the msg execution
+		panicTx      bool   // panic explicitly in tx execution
+		expErr       bool
+	}{
+		{"less than block gas meter", 10, false, false},
+		{"more than block gas meter", blockMaxGas, false, true},
+		{"more than block gas meter", uint64(float64(blockMaxGas) * 1.2), false, true},
+		{"consume block gas when paniced", 10, true, true},
```

Comment:
More tests to add:
1. We charge users `5000uatom` in gas but user only has `4000uatom` => transaction should fail and `4000uatom` must be charged.
2. Consume more then txLimit (not a block limit, but tx limit). Eg txLImit == 1000, we tx consumes 1200 => user should be charged 1200 (since the resources have been already used).


Probably we have more holes in the code.

Reply:
> More tests to add:
> 
> 1. We charge users `5000uatom` in gas but user only has `4000uatom` => transaction should fail and `4000uatom` must be charged.

If we fail to charge user the fee specified in tx, it won't execute the msg in the first place.

> 2. Consume more then txLimit (not a block limit, but tx limit). Eg txLImit == 1000, we tx consumes 1200 => user should be charged 1200 (since the resources have been already used).

`MempoolFeeMiddleware`'s checks in `checkTx` that: `txFee >= txGasLimit * minGasPrice`, and `DeductFeeMiddleware` just deduct `txFee` from payer. And it's reasonable since it'll panic immediately when tx gas limit exceeded.

> 
> Probably we have more holes in the code.

Reply:
> If we fail to charge user the fee specified in tx, we don't charge anything from user, and tx msgs are not executed neither.

This is wrong. We should charge user what's possible.

Reply:
Re 2.   This is about the gas consumed. This is not checked in `CheckTx`. We only know how much gas was consumed after processing a transaction.

Reply:
> > If we fail to charge user the fee specified in tx, we don't charge anything from user, and tx msgs are not executed neither.
> 
> This is wrong. We should charge user what's possible.

I was just describing what's currently implemented in cosmos-sdk, I think you proposed a new design on this, right?

> We charge users 5000uatom in gas but user only has 4000uatom => transaction should fail and 4000uatom must be charged.

If I understand correctly, the new design works like this:
```
if `user.balance >= txFee {
  deduct txFee from sender
} else {
  deduct user.balance from sender
  lower the tx gasLimit accordingly, (newGasLimit = user.balance / gasPrice)
  try to run tx msgs with the lowered gasLimit
}
```
There's one issue, what should be the `gasPrice` in the formula `newGasLimit = user.balance / gasPrice`, because we don't have `gasPrice` field in tx, the best thing we have is the `minGasPrice` in validator config, but that one could be empty.

Reply:
> If we fail to charge user the fee specified in tx, we don't charge anything from user, and tx msgs are not executed neither.

Yes, this is a change in the current functionality that is a good idea but out of scope of this PR. I would be happy if there was a new issue for this and it made it to main eventually. Since we fail fast in the ante handler, this has never been  viewed as an attack vector and I see no urgent need for this.

Reply:
OK, I will move it to the issue I created last week: https://github.com/cosmos/cosmos-sdk/issues/10908

Reply:
Yes, this is related to the issue I created last week: https://github.com/cosmos/cosmos-sdk/issues/10908, so we can continue there.